warehouse/query-core/src/main/java/datawave/query/ancestor/AncestorIndexIterator.java

```diff
@@ -1,32 +1,32 @@
 package datawave.query.ancestor;
 
-import com.google.common.base.Predicate;
-import com.google.common.base.Predicates;
-import datawave.query.iterator.filter.composite.CompositePredicateFilter;
 import datawave.query.iterator.logic.IndexIterator;
-import datawave.query.jexl.functions.FieldIndexAggregator;
-import datawave.query.jexl.functions.IdentityAggregator;
 import datawave.query.predicate.TimeFilter;
 import datawave.query.tld.TLD;
-import datawave.query.util.TypeMetadata;
 import org.apache.accumulo.core.data.Key;
 import org.apache.accumulo.core.data.Range;
 import org.apache.accumulo.core.data.Value;
 import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
 import org.apache.hadoop.io.Text;
 
-import java.util.Map;
-
 public class AncestorIndexIterator extends IndexIterator {
     
-    public AncestorIndexIterator(Text field, Text value, SortedKeyValueIterator<Key,Value> source, TimeFilter timeFilter) {
-        this(field, value, source, timeFilter, null, false, Predicates.<Key> alwaysTrue(), new IdentityAggregator(null, null), null);
+    public static class Builder<T extends AncestorIndexIterator,B extends Builder<T,B>> extends IndexIterator.Builder<AncestorIndexIterator,Builder<T,B>> {
+        protected Builder(Text field, Text value, SortedKeyValueIterator<Key,Value> source) {
+            super(field, value, source);
+        }
+        
+        public T build() {
+            return (T) new AncestorIndexIterator(this);
+        }
+    }
+    
+    public static Builder<?,?> builder(Text field, Text value, SortedKeyValueIterator<Key,Value> source) {
```

Comment:
You know things about the builder you are returning here. Can't you make the signature of this be something like the following?
```
public static <T extends AncestorIndexIterator,B extends Builder<T,B>> Builder<T,B> builder(...
```
If you return something that is qualified, the compiler can make assumptions about the return type, and I think you might be able to avoid having to do a cast everywhere you use this function.

Reply:
If I put this in IndexIterator:

```
    public static <T extends IndexIterator,B extends Builder<T,B>> Builder<T,B> builder(Text field, Text value, SortedKeyValueIterator<Key,Value> source) {
        return new Builder(field, value, source);
    }
```
then doing this in (the subclass) AncestorIndexIterator 

```
    public static <T extends AncestorIndexIterator,B extends Builder<T,B>> Builder<T,B> builder(Text field, Text value, SortedKeyValueIterator<Key,Value> source) {
        return new Builder(field, value, source);
    }

```


clashes and causes a compiler error:

`
[ERROR] /home/tom/projects/datawave/warehouse/query-core/src/main/java/datawave/query/ancestor/AncestorIndexIterator.java:[24,89] name clash: <T,B>builder(org.apache.hadoop.io.Text,org.apache.hadoop.io.Text,org.apache.accumulo.core.iterators.SortedKeyValueIterator<org.apache.accumulo.core.data.Key,org.apache.accumulo.core.data.Value>) in datawave.query.ancestor.AncestorIndexIterator and <T,B>builder(org.apache.hadoop.io.Text,org.apache.hadoop.io.Text,org.apache.accumulo.core.iterators.SortedKeyValueIterator<org.apache.accumulo.core.data.Key,org.apache.accumulo.core.data.Value>) in datawave.query.iterator.logic.IndexIterator have the same erasure, yet neither hides the other
`

The way they are now, there is no casting required when they are used.... you may have seen this 

```
        // @TODO: remove redundant type cast to check builder is building the subtype
        return (AncestorIndexIterator) AncestorIndexIterator.builder(field, value, source).withTimeFilter(timeFilter).withTypeMetadata(typeMetadata)
                        .shouldBuildDocument(buildDocument).withDatatypeFilter(datatypeFilter).withAggregation(aggregator).build();

```



and not realized that the typecast is unnecessary and is there temporarily only to make sure I don't break it while trying out various things. I plan to remove the typecasts (and instead make a unit test to confirm that the builder syntax is solid.

Reply:
Yeah, it seems the self-referential generics really complicate things. The static builder method adds confusion for someone trying to read the code and buys only a small readability improvement. You might consider removing it (and use new Foo.Builder()... instead), though I don't really have a strong opinion about it. Another couple nits... change if you want:
- The build methods could return the concrete type rather than a B
- You could add a "self()" method that does the cast of this to a B and avoid having to suppress or live with warnings on every "with*" method.